### PR TITLE
Results csv

### PR DESCRIPTION
**Pull request by goesslfabian**
Returns results for clause- and message-level evaluations as csv-files.

- clause-level: Stores the already created sklearn-classification-report in a csv. We should test its values against our own accuracy, precision, recall and f1 function calls. Note: On the clause-level, sklearn either returns the accuracy or the micro-averager, but never both. What is returned depends on whether the set of target labels in the val/test set exactly matches the labels that were seen during training.
- message-level: Now uses the multi-label format of the sklearn-classification-report to create the report.  We should check if we should also include a list of label-indices next to the target_names in the function call. This might not be necessary, as the mean-values of the classification report match the values from our own previous method of calculating the f1-score.

### CODE REVIEW

File: unify_eval/training/callback.py [hunk 1]
```diff
@@ -1025,6 +1046,95 @@ def __call__(self, model: Classifier, i_minibatch: int, iteration: int, *args, *
...
+        # Otherwise (if classes in pred/targets differ) sklearn returns the micro-average.
+        assert 'accuracy' in report or 'micro avg' in report
+
+        if 'accuracy' in report:
+            accuracy = report.pop('accuracy')
+        elif 'micro avg' in report:
+            accuracy = None
+        # this should be covered by the assert already
+        else:
+            return ValueError('Neither accuracy nor micro average available.')
```
Comment (marlon-betz) [inline unify_eval/training/callback.py:1064]: raise the exception, do not return it!

File: unify_eval/training/isolated_evaluation.py [hunk 2]
```diff
@@ -76,14 +86,78 @@ def __call__(self, model: Classifier, i_minibatch: int, iteration: int, *args, *
...
+        mlb.fit([tuple(self.labels_to_evaluate)])
+
+        # Create two matrices, one for predictions one for ground truth labels
+        y_pred_mlb = mlb.transform(y_pred_per_message)
+        y_true_mlb = mlb.transform(y_true_per_message)
+
+        # Calculate a classification report
+        # WARNING: this does not take the order of the predictions within the
+        # message into account but I think our legacy code does not either
+        # @ Reviewer. Let's talk about this.
```
Comment (marlon-betz) [inline unify_eval/training/isolated_evaluation.py:112]: positive, you can remove this comment

File: unify_eval/training/isolated_evaluation.py [hunk 3]
```diff
@@ -100,6 +174,14 @@ def __call__(self, model: Classifier, i_minibatch: int, iteration: int, *args, *
         self.writer.add_scalar(tag="isolated_precision", scalar_value=precision, global_step=self.global_step)
         self.writer.add_scalar(tag="isolated_recall", scalar_value=recall, global_step=self.global_step)
         self.writer.add_scalar(tag="isolated_f1", scalar_value=f1, global_step=self.global_step)
+
+        # Compare the results from the classification report to our legacy calculation method
+        # if used in production the legacy code should be removed
+        assert precision == df['MIC-precision'][0]
```
Comment (marlon-betz) [inline unify_eval/training/isolated_evaluation.py:180]: do these asserts work?

**Review by marlon-betz — CHANGES_REQUESTED** [review]

**Review by marlon-betz — CHANGES_REQUESTED** [review]
same here

File: unify_eval/training/isolated_evaluation.py [hunk 4]
```diff
@@ -76,14 +86,78 @@ def __call__(self, model: Classifier, i_minibatch: int, iteration: int, *args, *
...
 
                 counter += message_length
 
         self.data_loader.reset()
 
+        # Calculate classification report
+        # Fit a binary matrix where each column corresponds to a single label
+        # and each row to a single message
+        mlb = MultiLabelBinarizer()
+        mlb.fit([tuple(self.labels_to_evaluate)])
```
Comment (goesslfabian) [inline unify_eval/training/isolated_evaluation.py:103]: Turn self.labels_to_evaluate from a set into a list or tuple
